cpp/src/lists/explode.cu

```diff
@@ -208,41 +208,48 @@ std::unique_ptr<table> explode_outer(table_view const& input_table,
+      gather_map_p[index_to_write]             = lb_idx;
+      explode_col_gather_map_p[index_to_write] = idx;
+      if (include_position) {
+        position_array[index_to_write] = idx - (offsets[lb_idx] - offsets[0]);
+      }
+    }
+    if (null_or_empty[idx]) {
+      auto invalid_index = null_or_empty_offset_p[idx] == 0
+                             ? offsets[idx]
+                             : offsets[idx] + null_or_empty_offset_p[idx] - 1;
+      gather_map_p[invalid_index] = idx;
+
+      // negative one to indicate a null value
+      explode_col_gather_map_p[invalid_index] = -1;
+      if (include_position) { position_array[invalid_index] = 0; }
+    }
+  };
+
+  // we need to do this loop at least explode_col times or we may not properly fill in null and
+  // empty entries.
+  auto loop_count = std::max(sliced_child.size(), explode_col.size());
+
   // Fill in gather map with all the child column's entries
-  thrust::for_each(rmm::exec_policy(stream),
-                   counting_iter,
-                   counting_iter + sliced_child.size(),
-                   [offsets_minus_one,
-                    gather_map             = gather_map.begin(),
-                    explode_col_gather_map = explode_col_gather_map.begin(),
-                    position_array         = pos.begin(),
-                    include_position,
-                    offsets,
-                    null_or_empty_offset = null_or_empty_offset.begin(),
-                    null_or_empty,
-                    offset_size = explode_col.offsets().size() - 1] __device__(auto idx) {
-                     auto lb_idx = thrust::distance(
-                       offsets_minus_one,
-                       thrust::lower_bound(
-                         thrust::seq, offsets_minus_one, offsets_minus_one + (offset_size), idx));
-                     auto index_to_write                    = null_or_empty_offset[lb_idx] + idx;
-                     gather_map[index_to_write]             = lb_idx;
-                     explode_col_gather_map[index_to_write] = idx;
-                     if (include_position) {
-                       position_array[index_to_write] = idx - (offsets[lb_idx] - offsets[0]);
-                     }
-                     if (null_or_empty[idx]) {
-                       auto invalid_index = null_or_empty_offset[idx] == 0
-                                              ? offsets[idx]
-                                              : offsets[idx] + null_or_empty_offset[idx] - 1;
-                       gather_map[invalid_index] = idx;
-
-                       // negative one to indicate a null value
-                       explode_col_gather_map[invalid_index] = -1;
-
-                       if (include_position) { position_array[invalid_index] = 0; }
-                     }
-                   });
+  thrust::for_each(
+    rmm::exec_policy(stream), counting_iter, counting_iter + loop_count, fill_gather_maps);
```

Comment:
I'm confused about the `loop_count`. The two gather-maps are of size `sliced_child.size() + null_or_empty_count`. I don't see a loop in `fill_gather_maps()`, so wouldn't we need to iterate at least `gather_map_size()` times?

`loop_count` is currently `max(sliced_child.size(), explode_column.size())`. Consider:
```c++
auto explode_column = LCW{ {1,2,3}, {4,5}, {}, {}, {6} };
```
Shouldn't the explode column's gather map be of size `8`? How does this get filled if the `for_each` goes from `0` to `6` (exclusive)?

Reply:
P.S. Please pardon me if I've misunderstood the logic here.

Reply:
I was writing out the loop iterations here and I think you may be correct. I'm going to implement this as a test as I think it's great to test trailing nulls on the loop. I have a hunch that this test will fail and we do need to march the entire gather map.

Reply:
Ok, I've convinced myself that this loop is correct by writing tests and also using my whiteboard.

So the gather map creation is done by looping over `sliced_child.size()`. It is marching each row of the child data and looking at the offsets to determine the row that child data came from,  so we need to go `sliced_child.size()` times through the loop. 

The second part of the loop is filling in nulls and it is marching over the rows of the exploded column looking for duplicates. It also does this by looking at the offsets, but it's asking for each row if that row is null or empty. This needs to march over the `explode_column.size()`.